src/qt/importdialog.cpp

```diff
@@ -298,5 +344,377 @@ void ImportDialog::on_Accept_clicked() {
+                                            GetImportTimestamp(entry->getDescriptorData().timestamp, now),
+                                            scanned_time - TIMESTAMP_WINDOW - 1, TIMESTAMP_WINDOW);
+                                response.push_back(std::move(result));
+                            }
+                        }
+                    }
+                }
+            }
+            for (int i = 0; i < response.size(); i++) {
+                if (!response[i].success) {
+                    std::string warning_str;
+                    if (response[i].warnings.size() != 0) {
+                        for (auto w: response[i].warnings) {
+                            warning_str += w + "\n";
+                        }
+                    }
+                    QMessageBox::critical(this, QString::fromStdString("Error: input box # " + QString::number(i)),
+                                          QString::fromStdString("Error:" + response[i].error+"\n"+
+                                                                 "Warnings:"+"\n" + warning_str));
+                }
+            }
+        }
     }
 }
+
+
+
+ImportEntry *ImportDialog::addEntryMulti()
+{
+    ImportEntry *entry = new ImportEntry(ImportEntry::importMultiEntry, this);
+    entry->setModel(walletModel);
+    ui->entries->addWidget(entry);
+    connect(entry, &ImportEntry::removeMultiEntry, this, &ImportDialog::removeMultiEntry);
+
+    // Focus the field, so that entry can start immediately
+    //entry->clear();
+    entry->setFocus();
+    ui->scrollAreaWidgetContents->resize(ui->scrollAreaWidgetContents->sizeHint());
+    qApp->processEvents();
+    QScrollBar* bar = ui->scrollArea->verticalScrollBar();
+    if(bar)
+        bar->setSliderPosition(bar->maximum());
+
+    return entry;
+}
+
+ImportEntry *ImportDialog::addEntryDescriptors()
+{
+    ImportEntry *entry = new ImportEntry(ImportEntry::importDescriptorsEntry, this);
+    entry->setModel(walletModel);
+    ui->entries->addWidget(entry);
+    connect(entry, &ImportEntry::removeDescriptorEntry, this, &ImportDialog::removeDescriptorEntry);
+
+    // Focus the field, so that entry can start immediately
+    //entry->clear();
+    entry->setFocus();
+    ui->scrollAreaWidgetContents->resize(ui->scrollAreaWidgetContents->sizeHint());
+    qApp->processEvents();
+    QScrollBar* bar = ui->scrollArea->verticalScrollBar();
+    if(bar)
+        bar->setSliderPosition(bar->maximum());
+
+    return entry;
+}
+
+void ImportDialog::removeMultiEntry(ImportEntry* entry)
+{
+    entry->hide();
+
+    // If the last entry is about to be removed add an empty one
+    if (ui->entries->count() == 1)
+        addEntryMulti();
+
+    entry->deleteLater();
+}
+
+void ImportDialog::removeDescriptorEntry(ImportEntry* entry)
+{
+    entry->hide();
+
+    // If the last entry is about to be removed add an empty one
+    if (ui->entries->count() == 1)
+        addEntryDescriptors();
+
+    entry->deleteLater();
+}
```

Comment:
In c71f418f18000f537871f57a42ec76e0a25941e9 "qt: implemented importmulti and importdescriptors GUI's"

These two functions are basically the same. I think you could just collapse them into a single `removeEntry` function. To figure which `addEntry` to call, you can look at `entry`'s `entryPage` member (which will need to be made public, or a getter added).

Reply:
done